crates/in_stream/src/ws.rs

```diff
@@ -0,0 +1,412 @@
+                            Err(tungstenite::error::Error::Io(e)) if e.would_block() => {
+                                return Ok(())
+                            }
+                            Err(tungstenite::error::Error::Io(_)) => {
+                                if let Err(tungstenite::error::Error::Io(e)) = res {
+                                    return Err(e);
+                                } else {
+                                    unreachable!();
+                                }
+                            }
+                            Err(e) => {
+                                return Err(Error::new(
+                                    ErrorKind::Other,
+                                    format!("tungstenite error: {:?}", e),
+                                ))
+                            }
+                        }
+                    } else {
+                        return Ok(());
+                    }
+                }
+            }
+        }
+    }
+}
+
+impl<Sub: InStreamStd> InStream<&mut WsFrame, WsFrame> for InStreamWss<Sub> {
+    const URL_SCHEME: &'static str = SCHEME;
+
+    fn raw_connect<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = WssConnectConfig::from_gen(config)?;
+        validate_url_scheme(url)?;
+        let connect_url = url.clone();
+        let mut url = url.clone();
+        url.set_scheme(Sub::URL_SCHEME).unwrap();
+        let sub = Sub::raw_connect(&url, config.sub_connect_config)?;
+        let mut out = Self::priv_new(connect_url.clone());
+        match out.priv_proc_wss_cli_result(tungstenite::client(
+            tungstenite::handshake::client::Request {
+                url: connect_url.into(),
+                extra_headers: None,
+            },
+            sub.into_std_stream(),
+        )) {
+            Ok(_) => Ok(out),
+            Err(e) if e.would_block() => Ok(out),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn read(&mut self, data: &mut WsFrame) -> Result<usize> {
+        self.priv_process()?;
+        match &mut self.state {
+            None => Err(ErrorKind::NotConnected.into()),
+            Some(state) => match state {
+                WssState::Ready(wss) => match wss.read_message() {
+                    Ok(msg) => {
+                        data.assume(msg);
+                        Ok(1)
+                    }
+                    Err(tungstenite::error::Error::Io(e)) => Err(e),
+                    Err(e) => Err(Error::new(
+                        ErrorKind::Other,
+                        format!("tungstenite error: {:?}", e),
+                    )),
+                },
+                _ => Err(Error::with_would_block()),
+            },
+        }
+    }
+
+    fn write(&mut self, data: WsFrame) -> Result<usize> {
+        self.priv_process()?;
+        self.write_buf.push_back(data);
+        self.priv_write_pending()?;
+        Ok(1)
```

Comment:
Feels odd to me that this will always be one. Is this the consequence of forcing it to a frame and not having a `write_all(...)` batch equivalent?

Reply:
more like, this was me not wanting to introduce more generics to my traits...

Right now, we have R for the read type, W for the write type -- we could also add RR and WR for read-return and write-return so we could specify the return types as well.

The 1 to me *almost* makes sense... since the framing isn't an array type, yeah, you're just always dealing with individual elements.